### ISSUE TITLE
History : issue post 0.2.5 upgrade + [Feature Request]

### ISSUE THREAD

**Issue by WhatshallIbreaktoday** [PowerDNS-Admin/PowerDNS-Admin#1229]
Hi Guys, 

After upgrading to 0.2.5, I got an error on one of my servers (having 8k plus history log entries) :
2 other servers did not encounter the issue (~2K history logs)

Jun 24 11:24:55 myhost gunicorn[864132]: [2022-06-24 11:24:55,948] [app.py:1891] ERROR - Exception on /admin/history_table [GET]
Jun 24 11:24:55 myhost gunicorn[864132]: Traceback (most recent call last):
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
Jun 24 11:24:55 myhost gunicorn[864132]:     response = self.full_dispatch_request()
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
Jun 24 11:24:55 myhost gunicorn[864132]:     rv = self.handle_user_exception(e)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
Jun 24 11:24:55 myhost gunicorn[864132]:     reraise(exc_type, exc_value, tb)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
Jun 24 11:24:55 myhost gunicorn[864132]:     raise value
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
Jun 24 11:24:55 myhost gunicorn[864132]:     rv = self.dispatch_request()
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
Jun 24 11:24:55 myhost gunicorn[864132]:     return self.view_functions[rule.endpoint](**req.view_args)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/flask/lib/python3.8/site-packages/flask_login/utils.py", line 272, in decorated_view
Jun 24 11:24:55 myhost gunicorn[864132]:     return func(*args, **kwargs)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/powerdnsadmin/decorators.py", line 47, in decorated_function
Jun 24 11:24:55 myhost gunicorn[864132]:     return f(*args, **kwargs)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/powerdnsadmin/routes/admin.py", line 1238, in history_table
Jun 24 11:24:55 myhost gunicorn[864132]:     detailedHistories = convert_histories(histories)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/powerdnsadmin/routes/admin.py", line 915, in convert_histories
Jun 24 11:24:55 myhost gunicorn[864132]:     detailedHistories.append(DetailedHistory(histories[i], None))
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/opt/web/powerdns-admin/PowerDNS-Admin-0.2.5/powerdnsadmin/routes/admin.py", line 767, in __init__
Jun 24 11:24:55 myhost gunicorn[864132]:     detail_dict = json.loads(history.detail)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
Jun 24 11:24:55 myhost gunicorn[864132]:     return _default_decoder.decode(s)
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
Jun 24 11:24:55 myhost gunicorn[864132]:     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
Jun 24 11:24:55 myhost gunicorn[864132]:   File "/usr/lib/python3.8/json/decoder.py", line 353, in raw_decode
Jun 24 11:24:55 myhost gunicorn[864132]:     obj, end = self.scan_once(s, idx)
Jun 24 11:24:55 myhost gunicorn[864132]: json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

Maybe some old logs including wrong format.. or the amount of logs...     but the result is : blank page on history.. 
After erasing the logs, everything got back to normal... 

Just sharing, and also brings me a new [ feature request ]:  give the ability (to admins)  to dump the whole history to a  file before erasing them.  (in case of further auditing needs).

I still got a backup (mysql dump).. but it's less convenient :)

Great work anyway! this version rocks

**Comment by jbe-dw** [comment]
Hello,

I'm sorry you had this. We had not enough history to cover all possible outcomes.

Any feedback is welcome, thank you

**Comment by RGanor** [comment]
Is it possible for you to share your sqldump?

**Comment by mysidia** [comment]
I got a similar exception  with my history table from past Powerdnsadmin versions;   it seems that this module assumes entries in the  history table  contain valid JSON in the detail field.       However,    I had a few entries whose   detail value is at length  65535  characters and appears to be truncated,  so not valid JSON,

And many past history table entries referencing TXT records  contained JSON where quote symbol embedded in the TXT record seem to be not  escaped  such as

{"serial": "2019032201", "record": [{"record_name": "@", "record_type": "TXT", "record_status": "Active", "record_ttl": "600", "record_data": ""v=spf1 include:spf._____._______.___.com -all""}, {"record_name": "@", "record_type": "MX", "record_status": "Active", "record_ttl": "86400", "record_data": "10 smtpsnf.cust._______.___."}, {"record_name": "@", "record_type": "SOA", "record_status": "Active", "record_ttl": "7200", "record_data": "ns.____________.com. hostmaster.____________.com. 2019032201 28800 7200 604800 86400"}, {"record_name": "@", "record_type": "NS", "record_status": "Active", "record_ttl": "604800", "record_data": "ns.____________.com."}, {"record_name": "@", "record_type": "NS", "record_status": "Active", "record_ttl": "604800", "record_data": "ns2.____________.com."}, {"record_name": "mail", "record_type": "CNAME", "record_status": "Active", "record_ttl": "86400", "record_data": "maila._______.___."}, {"record_name": "maila", "record_type": "A", "record_status": "Active", "record_ttl": "86400", "record_data": "___.__.___.106"}, {"record_name": "mailtest", "record_type": "A", "record_status": "Active", "record_ttl": "300", "record_data": "___.__.___.108"}]}



MariaDB [powerdnsadmin]> select a.*,b.*,c.* from (select COUNT(*) AS count from
history) a, (select COUNT(*) AS invalid_count from history where json_valid(detail)=0) b,  (select COUNT(*) AS txt_related from history where json_valid(detail)
=0 and detail like '%"record_data": ""%""%') c;
+-------+---------------+-------------+
| count | invalid_count | txt_related |
+-------+---------------+-------------+
|  1174 |           913 |         909 |
+-------+---------------+-------------+
1 row in set (0.061 sec)





    detailedHistories.append(DetailedHistory(histories[i], None))
  File "/opt/web/powerdns-admin/./powerdnsadmin/routes/admin.py", line 767, in _
_init__
    detail_dict = json.loads(history.detail)
  File "/usr/lib/python3.9/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.9/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 146 (char 145)

**Comment by Finn10111** [comment]
I am having the same issue, after upgrading I could not login anymore. Thanks for the hint regarding the history table, I deleted all entries (about 2k) and now I am able to login again. If I can provide any further information please let me know, I would like to help.

**Comment by WhatshallIbreaktoday** [comment]
Hello, other "change" regarding history : 

Compared to the previous version, I cannot see detailed info for apikeys anymore :  
The line is present in database, also shown in history i.e
History page: 
Changed by | Content |Time |Detail
apikey |  delete zone NS record of test.domain.fr|2021-01-13 14:55:26|  ( No Info button shown here)

For a user, everything works fine.

**Comment by bynicolas** [comment]
Same issue regarding history not showing up in UI

I have this in the logs

```
gunicorn[5132]:     detailedHistories.append(DetailedHistory(histories[i], None))
gunicorn[5132]:   File "/var/www/html/pdns/powerdnsadmin/routes/admin.py", line 767, in __init__
gunicorn[5132]:     detail_dict = json.loads(history.detail)
gunicorn[5132]:   File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
gunicorn[5132]:     return _default_decoder.decode(s)
gunicorn[5132]:   File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
gunicorn[5132]:     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
gunicorn[5132]:   File "/usr/lib/python3.8/json/decoder.py", line 353, in raw_decode
gunicorn[5132]:     obj, end = self.scan_once(s, idx)
gunicorn[5132]: json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

**Comment by bynicolas** [comment]
After a bit more digging, the error I posted above stems from the `/dashboard` endpoint returning a `500` error.

I still can't see the history, but maybe these these 2 errors aren't related, I can't say, but seeing it complain about histories, I guess it does.

So the full error log, missing in the previous post
```
gunicorn[5132]: [2022-09-22 20:23:10,506] [app.py:1891] ERROR - Exception on /dashboard/ [GET]
gunicorn[5132]: Traceback (most recent call last):
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
gunicorn[5132]:     response = self.full_dispatch_request()
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
gunicorn[5132]:     rv = self.handle_user_exception(e)
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
gunicorn[5132]:     reraise(exc_type, exc_value, tb)
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
gunicorn[5132]:     raise value
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
gunicorn[5132]:     rv = self.dispatch_request()
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
gunicorn[5132]:     return self.view_functions[rule.endpoint](**req.view_args)
gunicorn[5132]:   File "/var/www/html/pdns/flask/lib/python3.8/site-packages/flask_login/utils.py", line 272, in decorated_view
gunicorn[5132]:     return func(*args, **kwargs)
gunicorn[5132]:   File "/var/www/html/pdns/powerdnsadmin/routes/dashboard.py", line 192, in dashboard
gunicorn[5132]:     detailedHistories = convert_histories(history)
gunicorn[5132]:   File "/var/www/html/pdns/powerdnsadmin/routes/admin.py", line 915, in convert_histories
gunicorn[5132]:     detailedHistories.append(DetailedHistory(histories[i], None))
gunicorn[5132]:   File "/var/www/html/pdns/powerdnsadmin/routes/admin.py", line 767, in __init__
gunicorn[5132]:     detail_dict = json.loads(history.detail)
gunicorn[5132]:   File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
gunicorn[5132]:     return _default_decoder.decode(s)
gunicorn[5132]:   File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
gunicorn[5132]:     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
gunicorn[5132]:   File "/usr/lib/python3.8/json/decoder.py", line 353, in raw_decode
gunicorn[5132]:     obj, end = self.scan_once(s, idx)
gunicorn[5132]: json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

**Comment by bynicolas** [comment]
So I managed to fix the double quote issue by applying this SQL Query on the history table

```
UPDATE history
SET detail = replace(detail, '\'', '"');
```

But I still have issues on the actual history page of my master server. The error is the same as @mysidia's, the `,` delimiter.

I can't seem to find the issue in the table and I don't know how to dump the contents of the affected variable to better understand what is going on

```
[2022-09-22 21:08:39,167] [app.py:1891] ERROR - Exception on /admin/history_table [GET]
gunicorn[1782]: Traceback (most recent call last):
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
gunicorn[1782]:     response = self.full_dispatch_request()
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
gunicorn[1782]:     rv = self.handle_user_exception(e)
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
gunicorn[1782]:     reraise(exc_type, exc_value, tb)
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
gunicorn[1782]:     raise value
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
gunicorn[1782]:     rv = self.dispatch_request()
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
gunicorn[1782]:     return self.view_functions[rule.endpoint](**req.view_args)
gunicorn[1782]:   File "/opt/web/powerdns-admin/venv/lib/python3.8/site-packages/flask_login/utils.py", line 272, in decorated_view
gunicorn[1782]:     return func(*args, **kwargs)
gunicorn[1782]:   File "/opt/web/powerdns-admin/powerdnsadmin/decorators.py", line 48, in decorated_function
gunicorn[1782]:     return f(*args, **kwargs)
gunicorn[1782]:   File "/opt/web/powerdns-admin/powerdnsadmin/routes/admin.py", line 1238, in history_table
gunicorn[1782]:     detailedHistories = convert_histories(histories)
gunicorn[1782]:   File "/opt/web/powerdns-admin/powerdnsadmin/routes/admin.py", line 915, in convert_histories
gunicorn[1782]:     detailedHistories.append(DetailedHistory(histories[i], None))
gunicorn[1782]:   File "/opt/web/powerdns-admin/powerdnsadmin/routes/admin.py", line 767, in __init__
gunicorn[1782]:     detail_dict = json.loads(history.detail)
gunicorn[1782]:   File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
gunicorn[1782]:     return _default_decoder.decode(s)
gunicorn[1782]:   File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
gunicorn[1782]:     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
gunicorn[1782]:   File "/usr/lib/python3.8/json/decoder.py", line 353, in raw_decode
gunicorn[1782]:     obj, end = self.scan_once(s, idx)
gunicorn[1782]: json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 97 (char 96)
```

**Comment by bynicolas** [comment]
@corubba Thanks for the snippet, it helped a lot narrowing down the issues.

It actually was 2 offending records containing badly escaped comments.

So I created an API key a while back and in the comments for that API key there was a single quote character. In the `history` table, that character wasn't escaped properly and was replaced by a double quote character which terminated prematurely the description attribute of the JSON object located in the `detail` column of the `history` table.

I was able to manually fix the issue, but I guess there is a bug in the code that is writing the history upon API key creation. Further inspection should be required to make sure the content is properly escaped.

**Comment by jbe-dw** [comment]
Hello,
The problem is that the way that detailed history records were set in the database was not pure json at first. We merged a PR that changed python calls to use json but it raised a lot of errors due to the poor format of the old logs.

We intended to reduce the impact by adding a database patch that replaced quotes (you had to run "flask db upgrade" from the deployment's virtualenv to apply the patch) but there was still some buggy records left, sadly not in any of the maintainers history, still causing the bug.

Maybe including @corubba 's snippet is a good idea.
